airflow/providers/amazon/aws/operators/appflow.py

```diff
@@ -0,0 +1,550 @@
+
+if TYPE_CHECKING:
+    from mypy_boto3_appflow.type_defs import (
+        DescribeFlowExecutionRecordsResponseTypeDef,
+        DescribeFlowResponseTypeDef,
+        ExecutionRecordTypeDef,
+        TaskTypeDef,
+    )
+
+    from airflow.utils.context import Context
+
+EVENTUAL_CONSISTENCY_OFFSET: int = 15  # seconds
+EVENTUAL_CONSISTENCY_POLLING: int = 10  # seconds
+SUPPORTED_SOURCES = {"salesforce", "zendesk"}
+MANDATORY_FILTER_DATE_MSG = "The filter_date argument is mandatory for {entity}!"
+NOT_SUPPORTED_SOURCE_MSG = "Source {source} is not supported for {entity}!"
+
+
+class AppflowBaseOperator(BaseOperator):
+    """
+    Amazon Appflow Base Operator class (not supposed to be used directly in DAGs).
+
+    :param source: The source name (Supported: salesforce, zendesk)
+    :param name: The flow name
+    :param flow_update: A boolean to enable/disable the a flow update before the run
+    :param source_field: The field name to apply filters
+    :param filter_date: The date value (or template) to be used in filters.
+    :param poll_interval: how often in seconds to check the query status
+    :param aws_conn_id: aws connection to use
+    :param region: aws region to use
+    """
+
+    BLUE = "#2bccbd"
+    ui_color = BLUE
+
+    def __init__(
+        self,
+        source: str,
+        name: str,
+        flow_update: bool,
+        source_field: Optional[str] = None,
+        filter_date: Optional[str] = None,
+        poll_interval: int = 20,
+        aws_conn_id: str = "aws_default",
+        region: Optional[str] = None,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        if source not in SUPPORTED_SOURCES:
+            raise ValueError(f"{source} is not a supported source (options: {SUPPORTED_SOURCES})!")
+        self.filter_date = filter_date
+        self.name = name
+        self.source = source
+        self.source_field = source_field
+        self.poll_interval = poll_interval
+        self.aws_conn_id = aws_conn_id
+        self.region = region
+        self.flow_update = flow_update
+
+    @cached_property
+    def hook(self) -> AppflowHook:
+        """Create and return an AppflowHook."""
+        return AppflowHook(aws_conn_id=self.aws_conn_id, region_name=self.region)
+
+    def execute(self, context: "Context") -> None:
+        self.filter_date_parsed: Optional[datetime] = (
+            datetime.fromisoformat(self.filter_date) if self.filter_date else None
+        )
+        if self.flow_update:
+            self._update_flow()
+        self._run_flow(context)
+
+    def _get_connector_type(self, response: "DescribeFlowResponseTypeDef") -> str:
+        connector_type = response["sourceFlowConfig"]["connectorType"]
+        if self.source != connector_type.lower():
+            raise ValueError(f"Incompatible source ({self.source} and connector type ({connector_type})!")
+        return connector_type
+
+    def _update_flow(self) -> None:
```

Comment:
To be honest, I think the logic for `_update_flow()` and `_run_flow()` should really live in the `AppflowHook` rather than an operator. Having these methods in the hook makes this integration more extensible for users and they won't be forced into a specific mode and sequence of interacting with AppFlow. Operators should, dare I say, "orchestrate" hook methods since hooks are meant to be the true interface and be more lightweight.

I'm certainly not saying to get rid of this base operator, but I think the base operator itself should use methods defined in the hook for the operationalized mode/sequence that's common for the other operators.

Reply:
It is a great suggestion indeed. I've just refactored the code to follow that strategy. 🙂